### PR TITLE
ci: enable playwright cache and parallel

### PR DESCRIPTION
**Pull request by himself65**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,7 +54,7 @@
     ]
   },
   "devDependencies": {
-    "@playwright/test": "^1.40.1",
+    "@playwright/test": "1.40.1",
```
Comment (dai-shi) [inline package.json:57]: Do we need this? Does it has to be in sync with ci.yml?

Comment (himself65) [reply package.json:57]: Yes. we can write another script to get the playwright version, let me try

Comment (himself65) [reply package.json:57]: Never mind, I don't wan't write another script for windows :(

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -38,9 +43,16 @@ jobs:
...
-      - run: pnpm run e2e
+      - name: Cache Playwright
+        uses: actions/cache@v3
+        id: playwright-cache
+        with:
+          path: ~/.cache/ms-playwright
+          key: ${{ runner.os }}-playwright-1.40.1
+      - name: Install Playwright
+        if: steps.playwright-cache.outputs.cache-hit != 'true'
+        run: pnpm dlx playwright@1.40.1 install --with-deps
```
Comment (dai-shi) [inline .github/workflows/ci.yml:54]: Just curious what happens with `pnpm playwright install --with-deps`.
It causes some errors, right?

I regularly update dependencies, so I wonder if I need to modify this when I update playwright in package.json.

Comment (himself65) [reply .github/workflows/ci.yml:54]: install --with-deps will install chrome, safari and firefox.

You will need to update this version too

Comment (dai-shi) [reply .github/workflows/ci.yml:54]:
```suggestion
        run: pnpm playwright install --with-deps
```
Let me try what will happen.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -43,15 +43,20 @@ jobs:
...
+          PLAYWRIGHT_VERSION=$(npm ls @playwright/test | grep -o '@playwright/test@[0-9]*\.[0-9]*\.[0-9]*' | awk -F@ '{print $3}')
+          echo "Playwright's Version: $PLAYWRIGHT_VERSION"
+          echo "PLAYWRIGHT_VERSION=$PLAYWRIGHT_VERSION" >> $GITHUB_ENV
       - name: Cache Playwright
         uses: actions/cache@v3
         id: playwright-cache
         with:
           path: ~/.cache/ms-playwright
-          key: ${{ runner.os }}-playwright-1.40.1
+          key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
```
Comment (dai-shi) [inline .github/workflows/ci.yml]: Oh, I see. You want to specify the version for the cache.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -43,15 +43,20 @@ jobs:
           cache-dependency-path: '**/pnpm-lock.yaml'
       - run: pnpm install --frozen-lockfile
       - run: pnpm run compile
+      - name: Store Playwright's Version
+        run: |
+          PLAYWRIGHT_VERSION=$(npm ls @playwright/test | grep -o '@playwright/test@[0-9]*\.[0-9]*\.[0-9]*' | awk -F@ '{print $3}')
```
Comment (dai-shi) [inline .github/workflows/ci.yml]: Does this work on windows?
`PLAYWRIGHT_VERSION=$(pnpm playwright --version)`

Maybe, it doesn't work with the space.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -38,9 +43,16 @@ jobs:
           cache-dependency-path: '**/pnpm-lock.yaml'
       - run: pnpm install --frozen-lockfile
       - run: pnpm run compile
-      - name: Install Playwright's dependencies
+      - name: Cache Playwright
+        uses: actions/cache@v3
+        id: playwright-cache
+        with:
+          path: ~/.cache/ms-playwright
+          key: ${{ runner.os }}-playwright-1.40.1
```
Comment (dai-shi) [inline .github/workflows/ci.yml:51]:
```suggestion
          key: ${{ runner.os }}-${{ hashFiles('pnpm-lock.yaml') }}
```
should be good. https://github.com/actions/cache

File: package.json (hunk 1 above)
Comment (dai-shi) [reply package.json:57]:
```suggestion
    "@playwright/test": "^1.40.1",
```

File: pnpm-lock.yaml [hunk 6]
```diff
@@ -12,7 +12,7 @@ importers:
   .:
     devDependencies:
       '@playwright/test':
-        specifier: ^1.40.1
+        specifier: 1.40.1
```
Comment (dai-shi) [inline pnpm-lock.yaml:15]:
```suggestion
        specifier: ^1.40.1
```

**Pull request merged**